### PR TITLE
Forward submitter exitcode to FlinkCluster status

### LINKED ISSUE #424: JobSubmitter exit code is lost
`Job.Mode: 'Blocking'`

- The jobsubmit script currently swallows the exit code of the main entrypoint [here](https://github.com/spotify/flink-on-k8s-operator/blob/master/controllers/flinkcluster/flinkcluster_submit_job_script.go#L156)

- Currently, the jobsubmitter is observed only while submitting the FlinkJob. The exit code for the submitter should also be observed and be potentially exposed in `FlinkCluster.Status.Components.Job` as `SubmitterExitCode`

```
status:
  components:
    job:
      completionTime: "2022-06-13T14:04:04Z"
      deployTime: "2022-06-13T11:19:14Z"
      id: somerandomId1234
      name: jobName1234
      startTime: "2022-06-13T11:32:02Z"
      state: Succeeded
      submitterName: flinkClusterName-job-submitter
      submitterExitCode: 0                                         <--- New field
``` 

`submitterExitCode` `-1` would mean that the submitter is still running. 
The downstream applications could potentially make use of this exitCode for retrying only a certain set of exit codes etc.

### PR DESCRIPTION
**Pull request by live-wire**
Fixes #424 

- Job submitter script swallows the job exit code. It should just forward it.
- The exit code of the submitter should be made available in the FlinkCluster Status
- JobState should be marked as Failed if the submitter exit code is not zero.[ This commit](https://github.com/spotify/flink-on-k8s-operator/pull/425/commits/4eb671b49500a143092d088acead12b6e53772b1) tries to make this happen. Do we need this? 🤔

### CODE REVIEW

**Review by kassybas — APPROVED** [review]

File: controllers/flinkcluster/flinkcluster_submit_job_script.go [hunk 1]
```diff
@@ -168,9 +168,7 @@ function main() {
     fi
 
     echo -e "\n---------- Submitting job ----------"
-    if ! submit_job "$@"; then
-        exit 2
-    fi
+    submit_job "$@"
```
[the remark this answers is not in the document]
Comment (live-wire) [reply controllers/flinkcluster/flinkcluster_submit_job_script.go:171]: This works as an exit code if the function returns a number.
(only with `set -e` - which is the case at the top of our script).  

TIL: thanks to @kassybas  :D

[the remark this answers is not in the document]
Comment (live-wire) [reply controllers/flinkcluster/flinkcluster_submit_job_script.go:172]: Turns out, this caused the script to die on a `write_term_log_msg` line :(

Using a more explicit (but not as pretty) approach now. @kassybas @jto

**Comment by kocomic** [comment]
after update flink-operator. Reconciler stuck in updateIfChanged(keep requeue), because existing oldJobStatus does not have field submitterExitCode. NewJobStatus is not deep equal to oldJobStatus, observer does not update observedCluster.status.

**Comment by kocomic** [comment]
Update flink-operator should consider compatible to existing old  job flinkclusters, status field breaking change should affect existing clusters state.

**Comment by live-wire** [comment]
@kocomic Thank you for reporting this 🙏  I'm on it.

**Comment by live-wire** [comment]
Hey @kocomic 

I think the submitterExitCode should at least take a default value `0` for the oldJobStatus. Nevertheless, the job status logic needs fixing.

Could you please share your `"Status changed"` logs? Just to be sure about where the problem is.


WIP PR: https://github.com/spotify/flink-on-k8s-operator/pull/427

**Comment by kocomic** [comment]
@live-wire updating logs
```json
{"level":"info","ts":1655984922.5693552,"logger":"controllers.FlinkCluster","msg":"---------- 2. Update cluster status ----------","cluster":"development/ad-event-generator"}
{
  level: 'info',
  ts: 1655984922.5693731,
  logger: 'controllers.FlinkCluster',
  msg: 'Job status changed',
  cluster: 'development/ad-event-generator',
  current: {
    name: 'Flink Streaming Job',
    submitterName: 'ad-event-generator-job-submitter',
    state: 'Updating',
    fromSavepoint: 'flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9',
    savepointGeneration: 28,
    savepointLocation: 'flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9',
    savepointTime: '2022-06-23T11:47:34Z',
    finalSavepoint: true,
    deployTime: '2022-06-23T11:47:34Z'
  },
  new: {
    name: 'Flink Streaming Job',
    submitterName: 'ad-event-generator-job-submitter',
    submitterExitCode: -1,
    state: 'Updating',
    fromSavepoint: 'flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9',
    savepointGeneration: 28,
    savepointLocation: 'flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9',
    savepointTime: '2022-06-23T11:47:34Z',
    finalSavepoint: true,
    deployTime: '2022-06-23T11:47:34Z'
  }
}
{
  "level": "info",
  "ts": 1655984922.5693848,
  "logger": "controllers.FlinkCluster",
  "msg": "Status changed",
  "cluster": "development/ad-event-generator",
  "old": {
    "state": "Updating",
    "components": {
      "configMap": {
        "name": "ad-event-generator-configmap",
        "state": "Updating"
      },
      "jobManagerStatefulSet": {
        "name": "ad-event-generator-jobmanager",
        "state": "Updating"
      },
      "jobManagerService": {
        "name": "ad-event-generator-jobmanager",
        "state": "Updating"
      },
      "taskManagerStatefulSet": {
        "name": "ad-event-generator-taskmanager",
        "state": "Updating"
      },
      "taskManagerDeployment": {
        "name": "",
        "state": ""
      },
      "job": {
        "name": "Flink Streaming Job",
        "submitterName": "ad-event-generator-job-submitter",
        "state": "Updating",
        "fromSavepoint": "flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9",
        "savepointGeneration": 28,
        "savepointLocation": "flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9",
        "savepointTime": "2022-06-23T11:47:34Z",
        "finalSavepoint": true,
        "deployTime": "2022-06-23T11:47:34Z"
      }
    },
    "savepoint": {
      "jobID": "c6ddbabcaf642e3269cc0e4006632725",
      "triggerID": "1234fb33ec21be504d7742056964adf7",
      "triggerTime": "2022-06-23T11:47:13Z",
      "triggerReason": "update",
      "requestTime": "2022-06-23T11:47:13Z",
      "state": "Succeeded"
    },
    "revision": {
      "currentRevision": "ad-event-generator-5d876bd896-2",
      "nextRevision": "ad-event-generator-5d9644b4df-3"
    },
    "lastUpdateTime": "2022-06-23T11:48:42Z"
  },
  "new": {
    "state": "Updating",
    "components": {
      "configMap": {
        "name": "ad-event-generator-configmap",
        "state": "Updating"
      },
      "jobManagerStatefulSet": {
        "name": "ad-event-generator-jobmanager",
        "state": "Updating"
      },
      "jobManagerService": {
        "name": "ad-event-generator-jobmanager",
        "state": "Updating"
      },
      "taskManagerStatefulSet": {
        "name": "ad-event-generator-taskmanager",
        "state": "Updating"
      },
      "taskManagerDeployment": {
        "name": "",
        "state": ""
      },
      "job": {
        "name": "Flink Streaming Job",
        "submitterName": "ad-event-generator-job-submitter",
        "submitterExitCode": -1,
        "state": "Updating",
        "fromSavepoint": "flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9",
        "savepointGeneration": 28,
        "savepointLocation": "flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9",
        "savepointTime": "2022-06-23T11:47:34Z",
        "finalSavepoint": true,
        "deployTime": "2022-06-23T11:47:34Z"
      }
    },
    "savepoint": {
      "jobID": "c6ddbabcaf642e3269cc0e4006632725",
      "triggerID": "1234fb33ec21be504d7742056964adf7",
      "triggerTime": "2022-06-23T11:47:13Z",
      "triggerReason": "update",
      "requestTime": "2022-06-23T11:47:13Z",
      "state": "Succeeded"
    },
    "revision": {
      "currentRevision": "ad-event-generator-5d876bd896-2",
      "nextRevision": "ad-event-generator-5d9644b4df-3"
    }
  }
}
{
  "level": "info",
  "ts": 1655984922.5694168,
  "logger": "controllers.FlinkCluster",
  "msg": "(ClusterStatusUpdater) Updating cluster status",
  "cluster": "development/ad-event-generator",
  "clusterClone": {
    "kind": "FlinkCluster",
    "apiVersion": "flinkoperator.k8s.io/v1beta1",
    "metadata": {
      "name": "ad-event-generator",
      "namespace": "development",
      "uid": "6584fad2-f5b5-4cbb-9ff5-667e0205a72d",
      "resourceVersion": "891937577",
      "generation": 3,
      "creationTimestamp": "2022-06-22T09:26:00Z",
      "labels": {
        "app.kubernetes.io/instance": "ad-event-generator",
        "app.kubernetes.io/managed-by": "Helm",
        "app.kubernetes.io/name": "flink-cluster",
        "helm.sh/chart": "flink-cluster-0.2.5"
      },
      "annotations": {
        "meta.helm.sh/release-name": "ad-event-generator",
        "meta.helm.sh/release-namespace": "development"
      }
    },
    "spec": {
      "flinkVersion": "1.13.1",
      "image": {
        "name": "hub.mydockerimages.net/public/ad-event-generator:v0.0.33",
        "pullPolicy": "IfNotPresent"
      },
      "jobManager": {
        "replicas": 1,
        "accessScope": "Cluster",
        "ServiceLabels": {
          "app.kubernetes.io/instance": "ad-event-generator",
          "app.kubernetes.io/managed-by": "Helm",
          "app.kubernetes.io/name": "flink-cluster",
          "helm.sh/chart": "flink-cluster-0.2.5"
        },
        "ports": {
          "rpc": 6123,
          "blob": 6124,
          "query": 6125,
          "ui": 8081
        },
        "extraPorts": [
          {
            "name": "prom",
            "containerPort": 9249,
            "protocol": "TCP"
          }
        ],
        "resources": {
          "limits": {
            "cpu": "2048m",
            "memory": "1Gi"
          },
          "requests": {
            "cpu": "400m",
            "memory": "1Gi"
          }
        },
        "memoryOffHeapMin": "0",
        "memoryProcessRatio": 80,
        "securityContext": {
          "runAsUser": 9999,
          "runAsGroup": 9999
        },
        "podLabels": {
          "app.kubernetes.io/instance": "ad-event-generator",
          "app.kubernetes.io/name": "flink-cluster"
        },
        "livenessProbe": {
          "tcpSocket": {
            "port": 6123
          },
          "initialDelaySeconds": 5,
          "timeoutSeconds": 10,
          "periodSeconds": 60,
          "failureThreshold": 5
        },
        "readinessProbe": {
          "tcpSocket": {
            "port": 6123
          },
          "initialDelaySeconds": 5,
          "timeoutSeconds": 10,
          "periodSeconds": 5,
          "failureThreshold": 60
        }
      },
      "taskManager": {
        "deploymentType": "StatefulSet",
        "replicas": 2,
        "ports": {
          "data": 6121,
          "rpc": 6122,
          "query": 6125
        },
        "extraPorts": [
          {
            "name": "prom",
            "containerPort": 9249,
            "protocol": "TCP"
          }
        ],
        "resources": {
          "limits": {
            "cpu": "1024m",
            "memory": "2Gi"
          },
          "requests": {
            "cpu": "500m",
            "memory": "1Gi"
          }
        },
        "memoryOffHeapMin": "0",
        "memoryProcessRatio": 80,
        "volumes": [
          {
            "name": "cache-volume",
            "emptyDir": {}
          }
        ],
        "volumeMounts": [
          {
            "name": "cache-volume",
            "mountPath": "/cache"
          }
        ],
        "podLabels": {
          "app.kubernetes.io/instance": "ad-event-generator",
          "app.kubernetes.io/name": "flink-cluster"
        },
        "livenessProbe": {
          "tcpSocket": {
            "port": 6122
          },
          "initialDelaySeconds": 5,
          "timeoutSeconds": 10,
          "periodSeconds": 60,
          "failureThreshold": 5
        },
        "readinessProbe": {
          "tcpSocket": {
            "port": 6122
          },
          "initialDelaySeconds": 5,
          "timeoutSeconds": 10,
          "periodSeconds": 5,
          "failureThreshold": 60
        }
      },
      "job": {
        "jarFile": "/opt/flink/usrlib/ad-event-generator.jar",
        "className": "com.public.analytics.ad.AdEventGeneratorJob",
        "allowNonRestoredState": false,
        "savepointsDir": "flink-state/ad-event-generator",
        "takeSavepointOnUpdate": true,
        "maxStateAgeToRestoreSeconds": 300,
        "autoSavepointSeconds": 3600,
        "parallelism": 6,
        "noLoggingToStdout": false,
        "restartPolicy": "FromSavepointOnFailure",
        "cleanupPolicy": {
          "afterJobSucceeds": "KeepCluster",
          "afterJobFails": "KeepCluster",
          "afterJobCancelled": "DeleteCluster"
        },
        "podLabels": {
          "app.kubernetes.io/instance": "ad-event-generator",
          "app.kubernetes.io/name": "flink-cluster"
        },
        "resources": {
          "limits": {
            "cpu": "2",
            "memory": "2Gi"
          },
          "requests": {
            "cpu": "200m",
            "memory": "512Mi"
          }
        },
        "mode": "Detached"
      },
      "envVars": [
        {
          "name": "CHECKPOINTS_OSS_DIR",
          "value": "flink-state/ad-event-generator/checkpoints"
        },
        {
          "name": "CHECKPOINT_INTERVAL_MILLIS",
          "value": "10000"
        },
        {
          "name": "KAFKA_SOURCE_BOOTSTRAP_SERVERS",
          "value": "10.2.224.33:9092"
        },
        {
          "name": "KAFKA_SOURCE_EVENTS_TOPIC",
          "value": "public.server_events"
        },
        {
          "name": "KAFKA_SOURCE_USER_LOGS_TOPIC",
          "value": "public.server_user_logs"
        },
        {
          "name": "KAFKA_SOURCE_LOGS_TOPIC",
          "value": "public.server_logs"
        },
        {
          "name": "KAFKA_SOURCE_GROUP_ID",
          "value": "ads"
        },
        {
          "name": "KAFKA_SINK_BOOTSTRAP_SERVERS",
          "value": "192.168.176.212:9092"
        },
        {
          "name": "KAFKA_SINK_TOPIC",
          "value": "event_confluence"
        },
        {
          "name": "KUDU_MASTERS",
          "value": "emr-header-1:7051"
        },
        {
          "name": "TEST_COUNT",
          "value": "2"
        }
      ],
      "envFrom": [
        {
          "secretRef": {
            "name": "ad-event-generator-secret-test"
          }
        }
      ],
      "flinkProperties": {
        "fs.oss.credentials.provider": "com.aliyun.oss.common.auth.EnvironmentVariableCredentialsProvider",
        "fs.oss.endpoint": "*",
        "metrics.reporter.prom.class": "org.apache.flink.metrics.prometheus.PrometheusReporter",
        "restart-strategy": "fixed-delay",
        "restart-strategy.fixed-delay.attempts": "10",
        "state.checkpoints.num-retained": "8",
        "taskmanager.numberOfTaskSlots": "3"
      },
      "logConfig": {
        "log4j-console.properties": "# Allows this configuration to be modified at runtime. The file will be checked every 30 seconds.\nmonitorInterval=3600\n# This affects logging for both user code and Flink\nrootLogger.level = INFO\nrootLogger.appenderRef.console.ref = ConsoleAppender\n# Uncomment this if you want to _only_ change Flink's logging\n#logger.flink.name = org.apache.flink\n#logger.flink.level = INFO\n\n# The following lines keep the log level of common libraries/connectors on\n# log level INFO. The root logger does not override this. You have to manually\n# change the log levels here.\nlogger.akka.name = akka\nlogger.akka.level = INFO\nlogger.kafka.name= org.apache.kafka\nlogger.kafka.level = INFO\nlogger.hadoop.name = org.apache.hadoop\nlogger.hadoop.level = INFO\nlogger.zookeeper.name = org.apache.zookeeper\nlogger.zookeeper.level = INFO\nlogger.shaded_zookeeper.name = org.apache.flink.shaded.zookeeper3\nlogger.shaded_zookeeper.level = INFO\n\n# Log all infos to the console\nappender.console.name = ConsoleAppender\nappender.console.type = CONSOLE\nappender.console.layout.type = PatternLayout\nappender.console.layout.pattern = %d{yyyy-MM-dd HH:mm:ss,SSS} %-5p %-60c %x - %m%n\n\n# Suppress the irrelevant (wrong) warnings from the Netty channel handler\nlogger.netty.name = org.apache.flink.shaded.akka.org.jboss.netty.channel.DefaultChannelPipeline\nlogger.netty.level = OFF\n"
      },
      "recreateOnUpdate": true
    },
    "status": {
      "state": "Updating",
      "components": {
        "configMap": {
          "name": "ad-event-generator-configmap",
          "state": "Updating"
        },
        "jobManagerStatefulSet": {
          "name": "ad-event-generator-jobmanager",
          "state": "Updating"
        },
        "jobManagerService": {
          "name": "ad-event-generator-jobmanager",
          "state": "Updating"
        },
        "taskManagerStatefulSet": {
          "name": "ad-event-generator-taskmanager",
          "state": "Updating"
        },
        "taskManagerDeployment": {
          "name": "",
          "state": ""
        },
        "job": {
          "name": "Flink Streaming Job",
          "submitterName": "ad-event-generator-job-submitter",
          "submitterExitCode": -1,
          "state": "Updating",
          "fromSavepoint": "flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9",
          "savepointGeneration": 28,
          "savepointLocation": "flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9",
          "savepointTime": "2022-06-23T11:47:34Z",
          "finalSavepoint": true,
          "deployTime": "2022-06-23T11:47:34Z"
        }
      },
      "savepoint": {
        "jobID": "c6ddbabcaf642e3269cc0e4006632725",
        "triggerID": "1234fb33ec21be504d7742056964adf7",
        "triggerTime": "2022-06-23T11:47:13Z",
        "triggerReason": "update",
        "requestTime": "2022-06-23T11:47:13Z",
        "state": "Succeeded"
      },
      "revision": {
        "currentRevision": "ad-event-generator-5d876bd896-2",
        "nextRevision": "ad-event-generator-5d9644b4df-3"
      },
      "lastUpdateTime": "2022-06-23T11:48:42Z"
    }
  },
  "cluster.Status": {
    "state": "Updating",
    "components": {
      "configMap": {
        "name": "ad-event-generator-configmap",
        "state": "Updating"
      },
      "jobManagerStatefulSet": {
        "name": "ad-event-generator-jobmanager",
        "state": "Updating"
      },
      "jobManagerService": {
        "name": "ad-event-generator-jobmanager",
        "state": "Updating"
      },
      "taskManagerStatefulSet": {
        "name": "ad-event-generator-taskmanager",
        "state": "Updating"
      },
      "taskManagerDeployment": {
        "name": "",
        "state": ""
      },
      "job": {
        "name": "Flink Streaming Job",
        "submitterName": "ad-event-generator-job-submitter",
        "submitterExitCode": -1,
        "state": "Updating",
        "fromSavepoint": "flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9",
        "savepointGeneration": 28,
        "savepointLocation": "flink-state/ad-event-generator/savepoint-c6ddba-5ae936606af9",
        "savepointTime": "2022-06-23T11:47:34Z",
        "finalSavepoint": true,
        "deployTime": "2022-06-23T11:47:34Z"
      }
    },
    "savepoint": {
      "jobID": "c6ddbabcaf642e3269cc0e4006632725",
      "triggerID": "1234fb33ec21be504d7742056964adf7",
      "triggerTime": "2022-06-23T11:47:13Z",
      "triggerReason": "update",
      "requestTime": "2022-06-23T11:47:13Z",
      "state": "Succeeded"
    },
    "revision": {
      "currentRevision": "ad-event-generator-5d876bd896-2",
      "nextRevision": "ad-event-generator-5d9644b4df-3"
    },
    "lastUpdateTime": "2022-06-23T11:48:42Z"
  }
}
{
  "level": "info",
  "ts": 1655984922.5834277,
  "logger": "controllers.FlinkCluster",
  "msg": "Wait status to be stable before taking further actions.",
  "cluster": "development/ad-event-generator",
  "requeueAfter": 5
}
```

**Comment by live-wire** [comment]
Hey @kocomic we made a couple of fixes.
-Now if you update the operator to the latest version `v0.4.1-beta.8`, the existing FlinkClusters (that don't have `submitterExitCode` defined in their status and still have the jobSubmitter pod available) will undergo 1 update for the same. And they shouldn't get stuck in re-queue state.

Please try it out :)

**Pull request merged**